File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.cpp

class DSU {
public:

int comps;
vector<int> parents;
vector<int> ranks;

    DSU(int n) {
        comps = n;
        parents.resize(n);
        for (int i = 0; i < n; i++) parents[i] = i;
        ranks.resize(n, 1);
    }

    int find(int i) {
        if (parents[i] == i) return i;
        return parents[i] = find(parents[i]);
    }

    bool uni(int i, int j) {
        int p1 = find(i);
        int p2 = find(j);
        if (p1 != p2) {
            if (ranks[p1] > ranks[p2]) {
                parents[p2] = p1;
                ranks[p1] += ranks[p2];
            } else {
                parents[p1] = p2;
                ranks[p2] += ranks[p1];
            }
            comps--;
            return true;
        }
        return false;
    }
};

class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        DSU dsu(n);
        for (auto edge: edges) {
            dsu.uni(edge[0], edge[1]);
        }
        return dsu.comps;
    }
};

File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.py
class Solution:
    
    def find(self, i, parents):
        if parents[i] == i:
            return i
        parents[i] = self.find(parents[i], parents)
        return parents[i]

    def union(self, i, j, parents, ranks):

        par1 = self.find(i, parents)
        par2 = self.find(j, parents)

        if par1 == par2:
            return False
        if ranks[par1] > ranks[par2]:
            ranks[par1] += ranks[par2]
            parents[par2] = par1
        else:
            ranks[par2] += ranks[par1]
            parents[par1] = par2
        
        return True

    
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        
        ranks = [1] * n
        parents = [i for i in range(n)]

        for i,j in edges:
            if self.union(i, j, parents, ranks):
                n -= 1
        
        return n
